### PR TITLE
EventSource testing components

### PR DESCRIPTION
**Pull request by anurse**
As a reminder, this is a `noship` package.

Adds an event listener and set of "Assert" helpers to allow collecting events and verifying that the expected events were written.

It wasn't quite as trivial as I'd first thought because `EventSource` is process-global so you need something that will work when multiple tests are run in parallel. I've got something that should work, and we can tweak it as needed.

`EventAssert` isn't your traditional assert class in that it is a static factory that creates a kind of "builder". This is needed because you can't always just compare event data directly to expectations since there are often dynamic values like timestamps and durations for which you want to provide a predicate rather than an exact expected value. I think this is a relatively clean way to do it, but I'm not 100%, so I'm happy for feedback.

@rynowak I know you weren't sure if we needed this but when playing with adding tracing to DI I literally hit a place where I wrote completely incorrect events that seemed right but ended up being totally wrong when I tested them out. I'm certainly open to using our judgement as to when to **use** these helpers to make tests, but the helpers themselves have some value in some cases.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Testing/Tracing/CollectingEventListener.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                }
+            }
+
+            // Mark this async context as "active" (will be ignored if isolateAsync is false)
+            _activeAsyncContext.Value = true;
+        }
+
+        public static IReadOnlyList<EventWrittenEventArgs> CollectEvents(Action action, params string[] eventSourceNames) => CollectEvents(action, true, eventSourceNames);
+
+        public static IReadOnlyList<EventWrittenEventArgs> CollectEvents(Action action, bool isolateAsync, params string[] eventSourceNames)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Testing/Tracing/CollectingEventListener.cs]: Old code. Removing the `isolateAsync` property.

**Comment by anurse** [comment]
@pakrym had an alternate suggestion to use a `CollectionAttribute` in xunit to ensure that all tests with EventListeners run in the same test collection and thus in sequence (preventing multiple async contexts from firing events). I'm going to spike that out and give it a quick test.

**Comment by anurse** [comment]
:up: :date: More differenter isolation. I like it better and it still allows the flexibility of handling strange async contexts. The only limitation is that tests that verify EventSources will **always** be run in sequence across the entire test assembly, and can never be run in parallel. That's OK because of how limited this testing will be, and it's most appropriate for how EventSource works anyway.

File: src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Testing.Tracing
+{
+    // This collection attribute is what makes the "magic" happen. It forces xunit to run all tests that inherit from this
+    // base class sequentially, preventing conflicts (since EventSource/EventListener is a process-global concept).
+    [Collection("Microsoft.AspNetCore.Testing.Tracing.EventSourceTests")]
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestBase.cs]: Now when I think more about it, even if event source tests are ran sequentially enabled event sources might cause other tests to log more events

Comment (pakrym) [reply src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestBase.cs]: Event  source tests have to run sequentially and no other tests should be running at the same time.

Comment (anurse) [reply src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestBase.cs]: True... maybe the async local is the only reliable way to isolate them then.

Comment (anurse) [reply src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestBase.cs]: I still like the base class patterns I got in the second iteration though, so I'll preserve that.

File: src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestCollection.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+namespace Microsoft.AspNetCore.Testing.Tracing
+{
+    // This file has to be public and has to actually live INSIDE the test assembly. Hence it is a shared-source file.
+    // This file is NOT part of Microsoft.AspNetCore.Testing.Tracing's compilation sources, it is exported to NuGet brought in to each
+    // see xunit bug: [TODO]
+    [Xunit.CollectionDefinition("Microsoft.AspNetCore.Testing.Tracing.EventSourceTestCollection", DisableParallelization = true)]
+    public class EventSourceTestCollection
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestCollection.cs]: This class is would be included both in `Microsoft.AspNetCore.Testing.Tracing` and user assembly, would it cause an issue?

**Comment by pakrym** [comment]
Remove `.binlog`

File: src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestCollection.cs (hunk 3 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.Testing/Tracing/EventSourceTestCollection.cs]: It shouldn't unless you try to use it. The assembly name will be different. I'm going to look at remove it though, it's not necessary that it be compiled in.

**Comment by rynowak** [comment]
> some value

OK I'll humour you 😆 

> The only limitation is that tests that verify EventSources will always be run in sequence across the entire test assembly, and can never be run in parallel

That doesn't seem like it goes far enough right? The tests that don't verify the event sources may still call APIs that produce events. So you don't you really need to serialize **all** tests?

File: test/Microsoft.AspNetCore.Testing.Tests/CollectingEventListenerTest.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                await Task.Yield();
+                TestEventSource.Log.TestWithPayload(42, 4.2);
+                await Task.Yield();
+
+                var events = GetEvents();
+                EventAssert.Collection(events,
+                    EventAssert.Event(1, "Test", EventLevel.Informational),
+                    EventAssert.Event(2, "TestWithPayload", EventLevel.Verbose)
+                        .Payload("payload1", 42)
+                        .Payload("payload2", 4.2));
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Testing.Tests/CollectingEventListenerTest.cs]: I think this is nice DSL

Comment (rynowak) [reply test/Microsoft.AspNetCore.Testing.Tests/CollectingEventListenerTest.cs]: I would pay $59.99 for 1024kbps down/248 up if it's ADSL

**Comment by pakrym** [comment]
> That doesn't seem like it goes far enough right? The tests that don't verify the event sources may still call > APIs that produce events. So you don't you really need to serialize all tests?

Test collection is marked with `[Xunit.CollectionDefinition(EventSourceTestBase.CollectionName, DisableParallelization = true)]` it would cause it to run serially after all tests that were able to run in parallel ended.

**Comment by rynowak** [comment]
Oh neat. Well this seems like it will work then.

**Comment by anurse** [comment]
Any concerns with me merging this?

**Comment by rynowak** [comment]
No concerns here :shipit:

**Comment by anurse** [comment]
I'll take that as an Approval :)

**Pull request merged**